File: 백준/Gold/17298. 오큰수/오큰수.java
import java.io.*;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;

    public static void main(String[] args) throws IOException {
        int N = Integer.parseInt(br.readLine());
        int[] list = new int[N];
        Stack<Integer> stack = new Stack<>();

        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < N; i++) {
            list[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = 0; i < N; i++) {
            while (!stack.isEmpty() && list[stack.peek()] < list[i]) {
                list[stack.pop()] = list[i];
            }
            stack.push(i);
        }

        while (!stack.isEmpty()) {
            list[stack.pop()] = -1;
        }

        for (int i = 0; i < N; i++) {
            bw.write(list[i] + " ");
        }

        bw.close();
    }
}

File: 백준/Gold/17298. 오큰수/오큰수.py
import sys


def input(): return sys.stdin.readline().rstrip()


N = int(input())
A = list(map(int, input().split()))

stack = []
result = [-1] * N
for i in range(N):
    if not stack:
        stack.append(i)
    else:
        while stack and A[stack[-1]] < A[i]:
            result[stack.pop()] = A[i]
        stack.append(i)
print(*result)